src/ray/raylet/local_object_manager.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/local_object_manager.h"
+
+namespace ray {
+
+namespace raylet {
+
+void LocalObjectManager::PinObjects(const rpc::Address &owner_address,
+                                    const std::vector<ObjectID> &object_ids,
+                                    std::vector<std::unique_ptr<RayObject>> &&objects) {
+  for (size_t i = 0; i < object_ids.size(); i++) {
+    const auto &object_id = object_ids[i];
+    auto &object = objects[i];
+    if (object == nullptr) {
+      RAY_LOG(ERROR) << "Plasma object " << object_id
+                     << " was evicted before the raylet could pin it.";
+      continue;
+    }
+    RAY_LOG(DEBUG) << "Pinning object " << object_id;
+    pinned_objects_.emplace(object_id, std::move(object));
+
+    // Send a long-running RPC request to the owner for each object. When we get a
+    // response or the RPC fails (due to the owner crashing), unpin the object.
+    // TODO(edoakes): we should be batching these requests instead of sending one per
+    // pinned object.
+    rpc::WaitForObjectEvictionRequest wait_request;
+    wait_request.set_object_id(object_id.Binary());
+    wait_request.set_intended_worker_id(owner_address.worker_id());
+    auto owner_client = owner_client_pool_.GetOrConnect(owner_address);
+    owner_client->WaitForObjectEviction(
+        wait_request,
+        [this, object_id](Status status, const rpc::WaitForObjectEvictionReply &reply) {
+          if (!status.ok()) {
+            RAY_LOG(WARNING) << "Worker failed. Unpinning object " << object_id;
+          }
+          RAY_LOG(DEBUG) << "Unpinning object " << object_id;
```

Comment:
Can we take out these logic into a different private method like `FreePinnedObjects`? It was a little confusing to me that Pin objects method also unpin objects.